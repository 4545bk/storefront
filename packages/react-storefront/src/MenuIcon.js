import React, { PureComponent } from 'react'
import SvgIcon from '@material-ui/core/SvgIcon'
import Open from './assets/menu.svg'
import Close from '@material-ui/icons/Close'
import { TransitionGroup, CSSTransition } from 'react-transition-group'
import { withStyles } from '@material-ui/core'
import PropTypes from 'prop-types'

export const styles = theme => ({
  root: {
    position: 'relative',
    height: '24px',
    width: '24px',
    boxSizing: 'border-box',

    '& .rsf-menu-icon-enter': {
      opacity: 0.01,
      transform: 'scale(0) rotate(45deg)'
    },

    '& .rsf-menu-icon-enter.rsf-menu-icon-enter-active': {
      opacity: 1,
      transform: 'scale(1) rotate(0deg)',
      transition: 'all 250ms ease-out'
    },

    '& .rsf-menu-icon-exit': {
      opacity: 1,
      transform: 'scale(1) rotate(0deg)'
    },

    '& .rsf-menu-icon-exit.rsf-menu-icon-exit-active': {
      opacity: 0.01,
      transform: 'scale(0) rotate(45deg)',
      transition: 'all 250ms ease-in'
    }
  },

  icon: {
    color: theme.palette.action.active,
    position: 'absolute',
    top: 0,
    left: 0
  }
})

/**
 * A menu icon that animates transitions between open and closed states.
 */
@withStyles(styles, { name: 'RSFMenuIcon' })
export default class MenuIcon extends PureComponent {

  static propTypes = {
    /**
     * Set to true when the menu is open, otherwise false
     */
    open: PropTypes.bool,

    /**
     * The icon to display when the menu is closed
     */
    OpenIcon: PropTypes.oneOfType([ PropTypes.string, PropTypes.func ]),

    /**
     * The icon to display when the menu is open
     */
    CloseIcon: PropTypes.oneOfType([ PropTypes.string, PropTypes.func ])
  }

  static defaultProps = {
    open: false,
    OpenIcon: Open,
    CloseIcon: Close
  }

  cssProps = { classNames: 'rsf-menu-icon', timeout: 300 }

  render() {
    const { open, classes, OpenIcon, CloseIcon } = this.props

    return (
      <div className={classes.root}>
        <TransitionGroup>
          {!open ? (
            <CSSTransition {...this.cssProps} key="open">
              <SvgIcon className={classes.icon}>
                <OpenIcon />
              </SvgIcon>
            </CSSTransition>
          ) : (
            <CSSTransition {...this.cssProps} key="close">
              <SvgIcon className={classes.icon}>
                <CloseIcon />
              </SvgIcon>
            </CSSTransition>
          )}
        </TransitionGroup>
      </div>
    )
  }

}