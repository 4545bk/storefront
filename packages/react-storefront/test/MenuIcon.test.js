import React from 'react'
import { mount } from 'enzyme'
import ExpandIcon from '@material-ui/icons/Add'
import CollapseIcon from '@material-ui/icons/Remove'
import MenuIcon from '../src/MenuIcon'

describe('MenuIcon', () => {
  it('should show the open icon when closed', () => {
    expect(mount(<MenuIcon />)).toMatchSnapshot()
  })

  it('should show the close icon when opened', () => {
    expect(mount(<MenuIcon open />)).toMatchSnapshot()
  })

  it('should render a label', () => {
    expect(mount(<MenuIcon label />)).toMatchSnapshot()
  })
})
