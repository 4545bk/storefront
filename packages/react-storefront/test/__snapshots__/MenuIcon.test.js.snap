// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuIcon should render a label 1`] = `
<WithStyles(MenuIcon) label={true}>
  <MenuIcon label={true} classes={{...}} open={false}>
    <div className="RSFMenuIcon-root-1 RSFMenuIcon-withLabel-3">
      <div className="rsf-hamburger" type="button">
        <span className="rsf-hamburger-box">
          <span className="rsf-hamburger-inner" />
        </span>
      </div>
      <div className="RSFMenuIcon-label-4">
        menu
      </div>
    </div>
  </MenuIcon>
</WithStyles(MenuIcon)>
`;

exports[`MenuIcon should show the close icon when opened 1`] = `
<WithStyles(MenuIcon) open={true}>
  <MenuIcon open={true} classes={{...}} label={true}>
    <div className="RSFMenuIcon-root-1 RSFMenuIcon-active-2 RSFMenuIcon-withLabel-3">
      <div className="rsf-hamburger" type="button">
        <span className="rsf-hamburger-box">
          <span className="rsf-hamburger-inner" />
        </span>
      </div>
      <div className="RSFMenuIcon-label-4">
        menu
      </div>
    </div>
  </MenuIcon>
</WithStyles(MenuIcon)>
`;

exports[`MenuIcon should show the open icon when closed 1`] = `
<WithStyles(MenuIcon)>
  <MenuIcon classes={{...}} open={false} label={true}>
    <div className="RSFMenuIcon-root-1 RSFMenuIcon-withLabel-3">
      <div className="rsf-hamburger" type="button">
        <span className="rsf-hamburger-box">
          <span className="rsf-hamburger-inner" />
        </span>
      </div>
      <div className="RSFMenuIcon-label-4">
        menu
      </div>
    </div>
  </MenuIcon>
</WithStyles(MenuIcon)>
`;
